File: Components/WinForms/VB/LayoutMainDemo/Modules/Browser.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Drawing
Imports System.Data
Imports System.Text
Imports System.Windows.Forms
Imports DevExpress.Utils
Imports DevExpress.LookAndFeel

Namespace DevExpress.XtraLayout.Demos
	Public Partial Class Browser
		Inherits DevExpress.XtraLayout.Demos.TutorialControl
		Private needDecreaseSelectedPageIndex As Boolean = False
		Private defaultUrl As String = ""
		Private dxUrl As String = "http://www.devexpress.com/"
		Public Sub New()
            InitializeComponent()
            DisabledVerticalStyle()
		End Sub
        Protected Overrides Sub OnStyleChanged()
            MyBase.OnStyleChanged()
            DisabledVerticalStyle()
        End Sub
        Sub DisabledVerticalStyle()
            checkEdit2.Enabled = UserLookAndFeel.Default.ActiveStyle <> ActiveLookAndFeelStyle.Flat _
                And UserLookAndFeel.Default.ActiveStyle <> ActiveLookAndFeelStyle.Office2003 _
                And UserLookAndFeel.Default.ActiveStyle <> ActiveLookAndFeelStyle.UltraFlat
        End Sub
        Protected Sub SetUrlInSelectedTabPage(ByVal url As String)
            If url = "" OrElse url Is Nothing Then
                CType((CType(tabbedControlGroup1.SelectedTabPage(0), LayoutControlItem)).Control, WebBrowser).GoHome()
            Else
                CType((CType(tabbedControlGroup1.SelectedTabPage(0), LayoutControlItem)).Control, WebBrowser).Url = New Uri(url)
            End If
        End Sub
        Protected Sub UpdateUrl()
            Dim wasError As Boolean = False
            Try
                SetUrlInSelectedTabPage(CStr(urlEdit.EditValue))
            Catch
                Try
                    SetUrlInSelectedTabPage("http://" & CStr(urlEdit.EditValue))
                Catch
                    wasError = True
                End Try
            End Try
            If (Not wasError) Then
                tabbedControlGroup1.SelectedTabPage.Text = CStr(urlEdit.EditValue)
            End If
        End Sub
        Private Sub urlEdit_ButtonClick(ByVal sender As Object, ByVal e As DevExpress.XtraEditors.Controls.ButtonPressedEventArgs) Handles urlEdit.ButtonClick
            UpdateUrl()
        End Sub
        Private Sub layoutControl1_TabPageCloseButtonClick(ByVal sender As Object, ByVal e As DevExpress.XtraLayout.Utils.LayoutGroupEventArgs) Handles layoutControl1.TabPageCloseButtonClick
            If tabbedControlGroup1.TabPages.Count = 2 Then
                Return
            End If
            DestroyTab(e.Group)
        End Sub
        Protected Sub CheckCloseButtonStatus()
            If tabbedControlGroup1.TabPages.Count = 2 Then
                tabbedControlGroup1.TabPages(0).ShowTabPageCloseButton = False
            End If
            If tabbedControlGroup1.TabPages.Count > 2 Then
                tabbedControlGroup1.TabPages(0).ShowTabPageCloseButton = True
            End If
        End Sub
        Public Sub DestroyTab(ByVal group As LayoutGroup)
            tabbedControlGroup1.BeginUpdate()
            CType((CType(group(0), LayoutControlItem)).Control, WebBrowser).Dispose()
            group.Dispose()
            CheckCloseButtonStatus()
            tabbedControlGroup1.EndUpdate()
        End Sub
        Private tabIndexToSelect As Integer = -1
        Private Sub tabbedControlGroup1_SelectedPageChanging(ByVal sender As Object, ByVal e As DevExpress.XtraLayout.LayoutTabPageChangingEventArgs) Handles tabbedControlGroup1.SelectedPageChanging
            If e.Page Is tabbedControlGroup1.TabPages(tabbedControlGroup1.TabPages.Count - 1) Then
                If (Not tabbedControlGroup1.IsUpdateLocked) Then
                    CreateBrowserTab(defaultUrl)
                    e.Cancel = True
                Else
                    needDecreaseSelectedPageIndex = True
                End If
            End If
        End Sub
        Protected Sub AsyncSelectTab()
            tabbedControlGroup1.SelectedTabPageIndex = tabIndexToSelect
        End Sub
        Private Sub tabbedControlGroup1_SelectedPageChanged(ByVal sender As Object, ByVal e As LayoutTabPageChangedEventArgs) Handles tabbedControlGroup1.SelectedPageChanged
            If needDecreaseSelectedPageIndex Then
                needDecreaseSelectedPageIndex = False
                tabbedControlGroup1.SelectedTabPageIndex -= 1
            End If
        End Sub
        Protected Sub CreateBrowserTab(ByVal url As String)
            layoutControl1.BeginUpdate()
            Dim newPage As LayoutControlGroup = New LayoutControlGroup()
            tabbedControlGroup1.InsertTabPage(tabbedControlGroup1.TabPages.Count - 1, newPage)
            newPage.ShowTabPageCloseButton = True
            Dim lci As LayoutControlItem = newPage.AddItem()
            lci.Padding = New XtraLayout.Utils.Padding(0)
            lci.TextVisible = False
            lci.BeginInit()
            Dim browser As WebBrowser = New WebBrowser()
            browser.Bounds = Rectangle.Empty 'antiflicker :)
            browser.Name = Guid.NewGuid().ToString()
            browser.ScriptErrorsSuppressed = True
            lci.Control = browser
            lci.EndInit()
            SetUrlInSelectedTabPage(url)
            If url <> "" Then
                newPage.Text = url
            Else
                newPage.Text = "New Page"
            End If
            CheckCloseButtonStatus()
            tabIndexToSelect = tabbedControlGroup1.TabPages.IndexOf(newPage)
            BeginInvoke(New MethodInvoker(AddressOf AsyncSelectTab))
            layoutControl1.EndUpdate()
        End Sub
        Private Sub UserControl1_Load(ByVal sender As Object, ByVal e As EventArgs) Handles MyBase.Load
            CreateBrowserTab(dxUrl)
            FillComboBoxValues()
        End Sub
        Protected Sub FillComboBoxValues()
            Dim names As String() = System.Enum.GetNames(tabbedControlGroup1.TextLocation.GetType())
            For Each name As String In names
                comboBoxEdit1.Properties.Items.Add(name)
            Next name
            comboBoxEdit1.SelectedIndex = 0
        End Sub
        Private Sub simpleButton1_Click(ByVal sender As Object, ByVal e As EventArgs)
            tabbedControlGroup1.BeginUpdate()
            tabbedControlGroup1.SelectedTabPage.HideToCustomization()
            tabbedControlGroup1.EndUpdate()
        End Sub

        Private Sub urlEdit_KeyDown(ByVal sender As Object, ByVal e As KeyEventArgs) Handles urlEdit.KeyDown
            If e.KeyCode = Keys.Enter Then
                UpdateUrl()
            End If
        End Sub

        Private Sub checkEdit1_CheckedChanged(ByVal sender As Object, ByVal e As EventArgs) Handles checkEdit1.CheckedChanged
            If checkEdit1.Checked Then
                tabbedControlGroup1.MultiLine = DefaultBoolean.True
            Else
                tabbedControlGroup1.MultiLine = DefaultBoolean.False
            End If
        End Sub

        Private Sub checkEdit2_CheckedChanged(ByVal sender As Object, ByVal e As EventArgs) Handles checkEdit2.CheckedChanged
            If checkEdit2.Checked Then
                tabbedControlGroup1.HeaderOrientation = DevExpress.XtraTab.TabOrientation.Vertical
            Else
                tabbedControlGroup1.HeaderOrientation = DevExpress.XtraTab.TabOrientation.Horizontal
            End If
        End Sub

        Private Sub comboBoxEdit1_SelectedIndexChanged(ByVal sender As Object, ByVal e As EventArgs) Handles comboBoxEdit1.SelectedIndexChanged
            tabbedControlGroup1.TextLocation = CType(System.Enum.Parse(GetType(Locations), CStr(comboBoxEdit1.Properties.Items(comboBoxEdit1.SelectedIndex))), Locations)
            checkEdit2_CheckedChanged(sender, e)
        End Sub
    End Class
End Namespace

File: Components/WinForms/CS/LayoutMainDemo/Modules/Browser.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Drawing;
using System.Data;
using System.Text;
using System.Windows.Forms;
using DevExpress.Utils;
using DevExpress.LookAndFeel;

namespace DevExpress.XtraLayout.Demos {
    public partial class Browser : DevExpress.XtraLayout.Demos.TutorialControl {
        bool needDecreaseSelectedPageIndex = false;
        string defaultUrl = "";
        string dxUrl = "http://www.devexpress.com/";
        public Browser() {
            InitializeComponent();
            DisabledVerticalStyle();
        }
        protected override void OnStyleChanged() {
            base.OnStyleChanged();
            DisabledVerticalStyle();
        }
        void DisabledVerticalStyle() {
            checkEdit2.Enabled = UserLookAndFeel.Default.ActiveStyle != ActiveLookAndFeelStyle.Flat 
                && UserLookAndFeel.Default.ActiveStyle != ActiveLookAndFeelStyle.Office2003
                && UserLookAndFeel.Default.ActiveStyle != ActiveLookAndFeelStyle.UltraFlat;
        }
        protected void SetUrlInSelectedTabPage(string url) {
            if (url == "" || url == null) ((WebBrowser)((LayoutControlItem)tabbedControlGroup1.SelectedTabPage[0]).Control).GoHome();
            else ((WebBrowser)((LayoutControlItem)tabbedControlGroup1.SelectedTabPage[0]).Control).Url = new Uri(url);
        }
        protected void UpdateUrl() {
            bool wasError = false;
            try {
                SetUrlInSelectedTabPage((string)urlEdit.EditValue);
            } catch {
                try {
                    SetUrlInSelectedTabPage("http://" + (string)urlEdit.EditValue);
                } catch {
                    wasError = true;
                }
            }
            if (!wasError) {
                tabbedControlGroup1.SelectedTabPage.Text = (string)urlEdit.EditValue;
            }
        }
        private void urlEdit_ButtonClick(object sender, DevExpress.XtraEditors.Controls.ButtonPressedEventArgs e) {
            UpdateUrl();
        }
        private void layoutControl1_TabPageCloseButtonClick(object sender, DevExpress.XtraLayout.Utils.LayoutGroupEventArgs e) {
            if (tabbedControlGroup1.TabPages.Count == 2) return;
            DestroyTab(e.Group);
        }
        protected void CheckCloseButtonStatus() {
            if (tabbedControlGroup1.TabPages.Count == 2) tabbedControlGroup1.TabPages[0].ShowTabPageCloseButton = false;
            if (tabbedControlGroup1.TabPages.Count > 2) tabbedControlGroup1.TabPages[0].ShowTabPageCloseButton = true;
        }
        public void DestroyTab(LayoutGroup group) {
            tabbedControlGroup1.BeginUpdate();
            ((WebBrowser)((LayoutControlItem)group[0]).Control).Dispose();
            group.Dispose();
            CheckCloseButtonStatus();
            tabbedControlGroup1.EndUpdate();
        }
        int tabIndexToSelect = -1;
        private void tabbedControlGroup1_SelectedPageChanging(object sender, DevExpress.XtraLayout.LayoutTabPageChangingEventArgs e) {
            if (e.Page == tabbedControlGroup1.TabPages[tabbedControlGroup1.TabPages.Count - 1]) {
                if (!tabbedControlGroup1.IsUpdateLocked) {
                    CreateBrowserTab(defaultUrl);
                    e.Cancel = true;
                } else {
                    needDecreaseSelectedPageIndex = true;
                }
            }
        }
        protected void AsyncSelectTab() {
            tabbedControlGroup1.SelectedTabPageIndex = tabIndexToSelect;
        }
        private void tabbedControlGroup1_SelectedPageChanged(object sender, LayoutTabPageChangedEventArgs e) {
            if (needDecreaseSelectedPageIndex) {
                needDecreaseSelectedPageIndex = false;
                tabbedControlGroup1.SelectedTabPageIndex--;
            }
        }
        protected void CreateBrowserTab(String url) {
            layoutControl1.BeginUpdate();
            LayoutControlGroup newPage = new LayoutControlGroup();
            tabbedControlGroup1.InsertTabPage(tabbedControlGroup1.TabPages.Count - 1, newPage);
            newPage.ShowTabPageCloseButton = true;
            LayoutControlItem lci = newPage.AddItem();
            lci.Padding = new DevExpress.XtraLayout.Utils.Padding(0);
            lci.TextVisible = false;
            lci.BeginInit();
            WebBrowser browser = new WebBrowser();
            browser.Bounds = Rectangle.Empty;//antiflicker :)
            browser.Name = Guid.NewGuid().ToString();
            browser.ScriptErrorsSuppressed = true;
            lci.Control = browser;
            lci.EndInit();
            SetUrlInSelectedTabPage(url);
            newPage.Text = url != "" ? url : "New Page";
            CheckCloseButtonStatus();
            tabIndexToSelect = tabbedControlGroup1.TabPages.IndexOf(newPage);
            BeginInvoke(new MethodInvoker(AsyncSelectTab));
            layoutControl1.EndUpdate();
        }
        private void UserControl1_Load(object sender, EventArgs e) {
            CreateBrowserTab(dxUrl);
            FillComboBoxValues();
        }
        protected void FillComboBoxValues() {
            string[] names = Enum.GetNames(tabbedControlGroup1.TextLocation.GetType());
            foreach (string name in names)
                comboBoxEdit1.Properties.Items.Add(name);
            comboBoxEdit1.SelectedIndex = 0;
        }
        private void simpleButton1_Click(object sender, EventArgs e) {
            tabbedControlGroup1.BeginUpdate();
            tabbedControlGroup1.SelectedTabPage.HideToCustomization();
            tabbedControlGroup1.EndUpdate();
        }

        private void urlEdit_KeyDown(object sender, KeyEventArgs e) {
            if (e.KeyCode == Keys.Enter) UpdateUrl();
        }

        private void checkEdit1_CheckedChanged(object sender, EventArgs e) {
            tabbedControlGroup1.MultiLine = checkEdit1.Checked ? DefaultBoolean.True : DefaultBoolean.False;
        }

        private void checkEdit2_CheckedChanged(object sender, EventArgs e) {
            tabbedControlGroup1.HeaderOrientation = checkEdit2.Checked ? DevExpress.XtraTab.TabOrientation.Vertical : DevExpress.XtraTab.TabOrientation.Horizontal;
        }

        private void comboBoxEdit1_SelectedIndexChanged(object sender, EventArgs e) {
            tabbedControlGroup1.TextLocation = (Locations)Enum.Parse(typeof(Locations), (string)comboBoxEdit1.Properties.Items[comboBoxEdit1.SelectedIndex]);
            checkEdit2_CheckedChanged(sender, e);
        }
    }
}
